File: twin/twin/client/java/src/main/java/org/ebayopensource/twin/NameMappings.java
// [Twin] Copyright eBay Inc., Twin authors, and other contributors.
// This file is provided to you under the terms of the Apache License, Version 2.0.
// See LICENSE.txt and NOTICE.txt for license and copyright information.

package org.ebayopensource.twin;

import java.util.*;

import org.ebayopensource.twin.element.*;
import org.ebayopensource.twin.pattern.*;

@SuppressWarnings("unchecked")
public class NameMappings {
	private static final Class<?>[] CONTROL_PATTERNS = {
		Editable.class,
		Expandable.class,
		Selectable.class,
		SelectionContainer.class,
		Toggle.class,
		Transformable.class,
	};

	private static final Class<?>[] CONTROL_TYPES = {
		Button.class,
		CalendarControl.class,
		CheckBox.class,
		ComboBox.class,
		Custom.class,
		DataGrid.class,
		DataItem.class,
		Desktop.class,
		Document.class,
		Edit.class,
		Group.class,
		Header.class,
		HeaderItem.class,
		Hyperlink.class,
		Image.class,
		ListControl.class,
		ListItem.class,
		Menu.class,
		MenuBar.class,
		MenuItem.class,
		Pane.class,
		ProgressBar.class,
		RadioButton.class,
		ScrollBarControl.class,
		Separator.class,
		Slider.class,
		Spinner.class,
		SplitButton.class,
		StatusBar.class,
		Tab.class,
		TabItem.class,
		Table.class,
		Text.class,
		Thumb.class,
		TitleBar.class,
		ToolBar.class,
		ToolTip.class,
		Tree.class,
		TreeItem.class,
		Window.class,
	};
	
	private static Map<String,Class<? extends ControlPattern>> patternNameToInterface = new HashMap<String,Class<? extends ControlPattern>>();
	private static Map<Class<? extends ControlPattern>,String> patternInterfaceToName = new HashMap<Class<? extends ControlPattern>,String>();
	private static Map<String,Class<? extends ControlType>> typeNameToInterface = new HashMap<String,Class<? extends ControlType>>();
	private static Map<Class<? extends ControlType>,String> typeInterfaceToName = new HashMap<Class<? extends ControlType>,String>();
	static {
		for(Class<?> pattern : CONTROL_PATTERNS) {
			Name nameAnnotation = pattern.getAnnotation(Name.class);
			String name = nameAnnotation == null ? pattern.getSimpleName() : nameAnnotation.value();
			if(!ControlPattern.class.isAssignableFrom(pattern))
				throw new IllegalStateException(pattern+" is not a ControlPattern");
			patternNameToInterface.put(name, (Class<? extends ControlPattern>)pattern);
			patternInterfaceToName.put((Class<? extends ControlPattern>)pattern, name);
		}
		for(Class<?> type : CONTROL_TYPES) {
			Name nameAnnotation = type.getAnnotation(Name.class);
			String name = nameAnnotation == null ? type.getSimpleName() : nameAnnotation.value();
			if(!ControlType.class.isAssignableFrom(type))
				throw new IllegalStateException(type+" is not a ControlPattern");
			typeNameToInterface.put(name, (Class<? extends ControlType>)type);
			typeInterfaceToName.put((Class<? extends ControlType>)type, name);
		}
	}
	public static Class<? extends ControlPattern> getPatternInterface(String name) {
		return patternNameToInterface.get(name);
	}
	public static String getPatternName(Class<? extends ControlPattern> iface) {
		return patternInterfaceToName.get(iface);
	}
	
	public static Class<? extends ControlType> getTypeInterface(String name) {
		return typeNameToInterface.get(name);
	}
	public static String getTypeName(Class<? extends ControlType> iface) {
		return typeInterfaceToName.get(iface);
	}
}

File: twin/twin/rc/src/main/csharp/NameMappings.cs
// [Twin] Copyright eBay Inc., Twin authors, and other contributors.
// This file is provided to you under the terms of the Apache License, Version 2.0.
// See LICENSE.txt and NOTICE.txt for license and copyright information.

using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Automation;

namespace Twin {
    class NameMappings {
        private static Dictionary<string, AutomationPattern> patterns = new Dictionary<string, AutomationPattern>();
        private static Dictionary<AutomationPattern, string> patternNames = new Dictionary<AutomationPattern, string>();
        private static Dictionary<AutomationPattern, AutomationProperty> patternPresenceProperties = new Dictionary<AutomationPattern, AutomationProperty>();

        private static Dictionary<string, ControlType> types = new Dictionary<string,ControlType>();
        private static Dictionary<ControlType, string> typeNames = new Dictionary<ControlType,string>();

        static NameMappings() {
            patterns["expand"] = ExpandCollapsePattern.Pattern;
            patterns["select"] = SelectionItemPattern.Pattern;
            patterns["select-container"] = SelectionPattern.Pattern;
            patterns["edit"] = ValuePattern.Pattern;
            patterns["toggle"] = TogglePattern.Pattern;
            patterns["transform"] = TransformPattern.Pattern;
            patternPresenceProperties[ExpandCollapsePattern.Pattern] = AutomationElement.IsExpandCollapsePatternAvailableProperty;
            patternPresenceProperties[SelectionItemPattern.Pattern] = AutomationElement.IsSelectionItemPatternAvailableProperty;
            patternPresenceProperties[SelectionPattern.Pattern] = AutomationElement.IsSelectionPatternAvailableProperty;
            patternPresenceProperties[ValuePattern.Pattern] = AutomationElement.IsValuePatternAvailableProperty;
            patternPresenceProperties[TogglePattern.Pattern] = AutomationElement.IsTogglePatternAvailableProperty;
            patternPresenceProperties[TransformPattern.Pattern] = AutomationElement.IsTransformPatternAvailableProperty;

            foreach (KeyValuePair<string, AutomationPattern> kv in patterns)
                patternNames[kv.Value] = kv.Key;

            types["Button"] = ControlType.Button;
            types["CalendarControl"] = ControlType.Calendar;
            types["CheckBox"] = ControlType.CheckBox;
            types["ComboBox"] = ControlType.ComboBox;
            types["Custom"] = ControlType.Custom;
            types["DataGrid"] = ControlType.DataGrid;
            types["DataItem"] = ControlType.DataItem;
            types["Document"] = ControlType.Document;
            types["Edit"] = ControlType.Edit;
            types["Group"] = ControlType.Group;
            types["Header"] = ControlType.Header;
            types["HeaderItem"] = ControlType.HeaderItem;
            types["Hyperlink"] = ControlType.Hyperlink;
            types["Image"] = ControlType.Image;
            types["ListControl"] = ControlType.List; // avoid List datatype 
            types["ListItem"] = ControlType.ListItem;
            types["Menu"] = ControlType.Menu;
            types["MenuItem"] = ControlType.MenuItem;
            types["MenuBar"] = ControlType.MenuBar;
            types["Pane"] = ControlType.Pane;
            types["ProgressBar"] = ControlType.ProgressBar;
            types["RadioButton"] = ControlType.RadioButton;
            types["ScrollBarControl"] = ControlType.ScrollBar; // avoid ScrollBar interface
            types["Separator"] = ControlType.Separator;
            types["Slider"] = ControlType.Slider;
            types["Spinner"] = ControlType.Spinner;
            types["SplitButton"] = ControlType.SplitButton;
            types["StatusBar"] = ControlType.StatusBar;
            types["Tab"] = ControlType.Tab;
            types["TabItem"] = ControlType.TabItem;
            types["Table"] = ControlType.Table;
            types["Text"] = ControlType.Text;
            types["Thumb"] = ControlType.Thumb;
            types["TitleBar"] = ControlType.TitleBar;
            types["ToolBar"] = ControlType.ToolBar;
            types["ToolTip"] = ControlType.ToolTip;
            types["Tree"] = ControlType.Tree;
            types["TreeItem"] = ControlType.TreeItem;
            types["Window"] = ControlType.Window;

            foreach (KeyValuePair<string, ControlType> kv in types)
                typeNames[kv.Value] = kv.Key;
        }

        public static AutomationPattern GetPattern(String name) {
            if (!patterns.ContainsKey(name))
                throw new Exception("Couldn't find AutomationPattern named " + name);
            return patterns[name];
        }
        public static AutomationProperty GetPatternPresenceProperty(AutomationPattern pattern) {
            if (!patternPresenceProperties.ContainsKey(pattern))
                throw new Exception("Couldn't find AutomationPattern presence property for " + pattern.ProgrammaticName);
            return patternPresenceProperties[pattern];
        }
        public static string GetName(AutomationPattern pattern) {
            if (!patternNames.ContainsKey(pattern))
                return null; // can return null if this is not a mapped pattern
            return patternNames[pattern];
        }

        public static ControlType GetType(String name) {
            if (!types.ContainsKey(name))
                throw new Exception("Couldn't find ControlType named " + name);
            return types[name];
        }
        public static string GetName(ControlType type) {
            if (!typeNames.ContainsKey(type))
                throw new Exception("Couldn't find ControlType name for " + type.ProgrammaticName);
            return typeNames[type];
        }
    }
}
